### PR TITLE
Fix/catch all invalid expression

### LINKED ISSUE #1000: apply_expression maps only KeyError to InvalidExpression, letting other exceptions escape
## Expected behavior and actual behavior.

I expected `rio_tiler.expression.apply_expression` to report an expression it cannot evaluate as `InvalidExpression`. It catches only `KeyError`, so every other failure escapes: `NotImplementedError`, `TypeError` and `OverflowError` from numexpr, and a `ValueError` from the `numpy.ma.MaskedArray(...)` construction in the same `try` block.

Issue #998 covered the expressions that validation can reject outright. The ones identified in this issue are the ones it cannot immediately reject. The bitwise nodes (`BitAnd`, `BitOr`, `BitXor`, `Invert`) work on integer arrays, so they cannot be dropped from `_ALLOWED_NODES`, and whether they evaluate depends on the array's dtype, which is not information in the expression itself. `apply_expression` must take into account the data's dtype.

A number of expressions pass validation and then fail in `apply_expression` with something other than `KeyError` (arrays are `numpy.ones(shape=(2, 10, 10))` of the dtype shown, see below for steps to reproduce):

```text
expression        dtype     escaping exception
b1 & b2           float32   NotImplementedError: couldn't find matching opcode for 'and_fff'
b1 | b2           float32   NotImplementedError: couldn't find matching opcode for 'or_fff'
b1 ^ b2           float32   NotImplementedError: couldn't find matching opcode for 'xor_fff'
~b1               float32   NotImplementedError: couldn't find matching opcode for 'invert_ff'
b1 & 1.5          int32     NotImplementedError: couldn't find matching opcode for 'and_did'
sqrt(b1, b2)      uint8     NotImplementedError: couldn't find matching opcode for 'sqrt_dii'
b1 < b2 < b1      uint8     TypeError: You can't use Python's standard boolean operators ...
where(b1, b2)     uint8     TypeError: where_func() missing 1 required positional argument: 'c'
b1 + "a"          uint8     TypeError: strings can only be operated with strings
b1 ** 1e400       uint8     OverflowError: cannot convert float infinity to integer
b1;1              uint8     ValueError: setting an array element with a sequence ...
```

Expected `InvalidExpression` in all cases.

## Steps to reproduce the problem.

To reproduce, add this test to `tests/test_expression.py`:

```python
@pytest.mark.parametrize(
    "expr,dtype",
    [
        # bitwise operators: valid on integer arrays, unsupported on floats
        ("b1 & b2", "float32"),
        ("b1 | b2", "float32"),
        ("b1 ^ b2", "float32"),
        ("~b1", "float32"),
        # bitwise operator with a float constant, on an integer array
        ("b1 & 1.5", "int32"),
        # chained comparison
        ("b1 < b2 < b1", "uint8"),
        # wrong number of arguments to an allowed function
        ("where(b1, b2)", "uint8"),
        ("sqrt(b1, b2)", "uint8"),
        # string operand
        ('b1 + "a"', "uint8"),
        # non-finite constant
        ("b1 ** 1e400", "uint8"),
        # block that evaluates to a scalar
        ("b1;1", "uint8"),
    ],
)
def test_apply_expression_invalid(expr, dtype):
    """apply_expression should report invalid expressions as InvalidExpression."""
    data = numpy.ma.MaskedArray(numpy.ones(shape=(2, 10, 10), dtype=dtype))

    with pytest.raises(InvalidExpression):
        apply_expression(get_expression_blocks(expr), ["b1", "b2"], data)
```

All fail against 9.4.5:

```text
FAILED tests/test_expression.py::test_apply_expression_invalid[b1 & b2-float32]
FAILED tests/test_expression.py::test_apply_expression_invalid[b1 | b2-float32]
FAILED tests/test_expression.py::test_apply_expression_invalid[b1 ^ b2-float32]
FAILED tests/test_expression.py::test_apply_expression_invalid[~b1-float32]
FAILED tests/test_expression.py::test_apply_expression_invalid[b1 & 1.5-int32]
FAILED tests/test_expression.py::test_apply_expression_invalid[b1 < b2 < b1-uint8]
FAILED tests/test_expression.py::test_apply_expression_invalid[where(b1, b2)-uint8]
FAILED tests/test_expression.py::test_apply_expression_invalid[sqrt(b1, b2)-uint8]
FAILED tests/test_expression.py::test_apply_expression_invalid[b1 + "a"-uint8]
FAILED tests/test_expression.py::test_apply_expression_invalid[b1 ** 1e400-uint8]
FAILED tests/test_expression.py::test_apply_expression_invalid[b1;1-uint8]
```

A possible fix: add a fallback clause to `apply_expression`, leaving the `KeyError` message as it is.

```diff
     except KeyError as e:
         raise InvalidExpression(f"Invalid band/asset name {str(e)}") from e
+    except MemoryError:
+        raise
+    except Exception as e:
+        raise InvalidExpression(f"Invalid expression: {e}") from e
```

With that, all pass and the rest of the suite is unchanged.

Note that `MemoryError` is re-raised because an allocation failure depends on the machine. However, the proposed new tests do not reproduce this, so the suggestion to single out `MemoryError` is a defensive suggestion, so that such cases do not produce a 400 response, since the client does not control the size of the machine.

We could explicitly handle the other error types, rather than using the `Exception` catch-all, but that would require tracking the various types of exceptions that numexpr and numpy could raise, and numexpr doesn't even document the specific exceptions it can raise.

#### Environment Information

- rio-tiler: 9.4.5
- Python version: 3.12.14 (main, Aug 14 2026, 15:24:41) [Clang 22.1.3 ]
- Operating System: macOS-26.6.2-arm64-arm-64bit
- numexpr 2.14.2
- numpy 2.5.1

## Installation Method

`uv sync` in a clone of the repository at `b7d4bc9` (9.4.5), for the test runs. Originally found with the 9.4.3 wheel from PyPI, installed by `uv` as a transitive dependency.

## AI tool usage

 - [x] AI (Copilot or something similar) supported my development of this issue. See our [policy about AI tool use](https://github.com/cogeotiff/.github/ai-contribution-policy.md). Use of AI tools *must* be indicated.

### PR DESCRIPTION
**Pull request by vincentsarago**
## Technical Context
- **Related Issues:** closes #1000 
- **Breaking Changes:** [Yes/No] (If yes, please describe the impact and migration path)

---

## Checklist
- [x] **Linting:** Code is formatted and linted
- [x] **Tests:** Tests pass. I have included new tests for these changes where applicable.
- [x] **Edge Cases:** I have manually verified "unhappy paths" and edge cases beyond the basic success criteria (e.g., database connection timeouts, malformed input, strict mapping rejections).
- [ ] **Documentation:** I have updated `README.md` to reflect any new environment variables, configuration changes, or breaking updates.
- [x] **Accountability:** I can explain the implementation logic for every line of code submitted.

---

## AI tool usage
 - [ ] AI (Copilot or something similar) supported my development of this PR. See our [policy about AI tool use](https://github.com/cogeotiff/.github/ai-contribution-policy.md). Use of AI tools *must* be indicated.
 ---
 > **Policy:** We require a "human-in-the-loop." You are the author and are fully accountable for all submitted code. Please ensure all tool-generated content is thoroughly reviewed before submission to ensure it is not an "extractive contribution" that squanders maintainer time.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -1,6 +1,10 @@
 
 # Unreleased
 
+## 9.4.5 (2026-09-17)
```
Comment (chuckwondo) [inline CHANGES.md:4]:
```suggestion
## 9.4.6 (2026-09-17)
```

Comment (vincentsarago) [reply CHANGES.md:4]: done in https://github.com/cogeotiff/rio-tiler/commit/95362365e29ec55ce1dcd972ebeefaa746fd36b5
